### PR TITLE
FSDP: use Work.wait instead of event for all reduce

### PR DESCRIPTION
**Pull request by d4l3k**
This uses Work.swait instead of using CUDA events for allreduce. This has some important performance considerations when using ProcessGroups that depend on CPU synchronization such as Gloo. By not calling `Work.wait` until the end this allows for all of the allreduce calls to be dispatched before we synchronize on any of them.

# Test plan

Run with torchtitan and verify memory and tps is the same

```
REPLICA_GROUP_ID=1 NGPU=4 CONFIG_FILE=./torchtitan/models/llama/train_configs/llama3_8b.toml ./run_train.sh --training.data_parallel_shard_degree=2 --training.data_parallel_replicate_degree=2  --metrics.log_freq=1 --training.seq_len 4096
```

## Baseline

```
[rank0]:[titan] 2025-03-06 16:50:02,110 - root - INFO - step:  4  loss: 11.3703  memory: 80.00GiB(84.18%)  tps: 454  tflops: 23.36  mfu: 2.36%
[rank0]:[titan] 2025-03-06 16:50:11,060 - root - INFO - step:  5  loss: 11.1204  memory: 80.00GiB(84.18%)  tps: 458  tflops: 23.56  mfu: 2.38%
[rank0]:[titan] 2025-03-06 16:50:19,816 - root - INFO - step:  6  loss: 10.5902  memory: 80.00GiB(84.18%)  tps: 468  tflops: 24.08  mfu: 2.44%
[rank0]:[titan] 2025-03-06 16:50:28,662 - root - INFO - step:  7  loss: 10.4288  memory: 80.00GiB(84.18%)  tps: 463  tflops: 23.84  mfu: 2.41%
[rank0]:[titan] 2025-03-06 16:50:37,136 - root - INFO - step:  8  loss: 10.6261  memory: 80.00GiB(84.18%)  tps: 484  tflops: 24.89  mfu: 2.52%
[rank0]:[titan] 2025-03-06 16:50:45,576 - root - INFO - step:  9  loss: 10.6391  memory: 80.00GiB(84.18%)  tps: 485  tflops: 24.98  mfu: 2.53%
[rank0]:[titan] 2025-03-06 16:50:54,317 - root - INFO - step: 10  loss: 10.3310  memory: 80.00GiB(84.18%)  tps: 469  tflops: 24.12  mfu: 2.44%
[rank0]:[titan] 2025-03-06 16:51:03,457 - root - INFO - step: 11  loss:  9.8705  memory: 80.00GiB(84.18%)  tps: 448  tflops: 23.07  mfu: 2.33%
```
avg = 466 tps, std = 12.4

## With Change

```
[rank0]:[titan] 2025-03-06 16:44:48,055 - root - INFO - step:  5  loss: 11.0233  memory: 80.00GiB(84.18%)  tps: 468  tflops: 24.08  mfu: 2.44%
[rank0]:[titan] 2025-03-06 16:44:56,778 - root - INFO - step:  6  loss: 10.5089  memory: 80.00GiB(84.18%)  tps: 470  tflops: 24.17  mfu: 2.44%
[rank0]:[titan] 2025-03-06 16:45:05,606 - root - INFO - step:  7  loss: 10.2253  memory: 80.00GiB(84.18%)  tps: 464  tflops: 23.89  mfu: 2.42%
[rank0]:[titan] 2025-03-06 16:45:14,203 - root - INFO - step:  8  loss: 10.5021  memory: 80.00GiB(84.18%)  tps: 477  tflops: 24.53  mfu: 2.48%
[rank0]:[titan] 2025-03-06 16:45:22,986 - root - INFO - step:  9  loss: 10.5685  memory: 80.00GiB(84.18%)  tps: 466  tflops: 24.01  mfu: 2.43%
[rank0]:[titan] 2025-03-06 16:45:31,795 - root - INFO - step: 10  loss: 10.4201  memory: 80.00GiB(84.18%)  tps: 465  tflops: 23.94  mfu: 2.42%
[rank0]:[titan] 2025-03-06 16:45:40,274 - root - INFO - step: 11  loss:  9.8893  memory: 80.00GiB(84.18%)  tps: 483  tflops: 24.87  mfu: 2.51%
```

avg = 470 tps, std = 6.5

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @c-p-i-o

### CODE REVIEW

**Comment by fegin** [comment]
@awgu Do you have any concerns about the change?

**Comment by awgu** [comment]
I am not sure if this change is correct. We need the FSDP all-reduce stream to wait on the all-reduce PG's internal stream if we ever run any subsequent ops after the all-reduce (e.g. upcast bf16 to fp32). Changing the all-reduce to use `async_op=True` break this.

**Comment by d4l3k** [comment]
@awgu can you share where the code pointer to that is? 

My reading of the code is that we never synchronize on the allreduce stream other than the 1 event I removed in this PR

There is the `all_reduce_hook` which also uses the stream (#147114) but that's not related to bf16?

**Comment by d4l3k** [comment]
Ahh I see the `post_reduce_stream` OK, yeah, I think you're right

**Comment by kwen2501** [comment]
You may need [this PR](https://github.com/pytorch/pytorch/pull/148590) to avoid peak memory usage increase. It is now landed.

**Comment by kwen2501** [comment]
In general, I feel that work has better abstraction than stream + event. Which is what I like about this PR.

Am I understanding it right that there is a dependency like this: reduce-scatter -> all-reduce -> post-reduction processing? If that's the case, can we possibly use one reduction stream to do all 3?

**Comment by awgu** [comment]
We want all-reduce to not run in the same stream as reduce-scatter so that all-reduce can overlap with later reduce-scatters (i.e. all-reduce are async wrt. reduce-scatter).

**Comment by kwen2501** [comment]
@awgu Sure, agree. Just a naive thought, can the latter reduce-scatter have its own stream? Said differently, can we arrange streams based on FSDP instances instead of comm op type? It seems the code can be easier if written that way.

**Comment by kwen2501** [comment]
image

**Comment by kwen2501** [comment]
(Continuing)
Today we arrange streams along the dotted lines, as a result, we are forced to write event sync explicitly to establish the solid lines. Would it be easier if we use stream's inherent dependency for the solid lines in the first place?

**Comment by d4l3k** [comment]
@kwen2501 would this require having one stream per parameter group? I'm also not sure this really solves the CPU synchronization piece since we would still block the CPU thread on sync -- the "optimal" way of handling the CPU side of things in a GPU+CPU compatible way would be to use PyTorch Futures

**Comment by kwen2501** [comment]
In case of GPU + FSDP, there should not be CPU sync in the above figure.

Are you talking about CPU + FSDP case? Then Yes, an instance-to-instance sync would be needed. But that's FSDP's **explicit choice**. In a world without memory restriction, I don't see why two reduce-scatters cannot happen simultaneously.

**Comment by kwen2501** [comment]
> would this require having one stream per parameter group?

Yes it will, and that's the point -- to keep various processings (rs, ar, post, etc) for one parameter group on the same stream, so that their dependency is naturally observed.

**Comment by albanD** [comment]
> In general, I feel that work has better abstraction than stream + event. Which is what I like about this PR.

I'm afraid that ship has sailed. Every other system speaks in stream events because that's what we have to use for raw cuda interactions. I don't think it is helpful to introduce another one since we cannot migrate everything to it.

> Are you talking about CPU + FSDP case?

What use case is that? I agree with @kwen2501 's comment above that on the regular FSDP case, there is no more CPU sync (which is the whole point). https://dev-discuss.pytorch.org/t/rethinking-pytorch-fully-sharded-data-parallel-fsdp-from-first-principles/1019 and https://dev-discuss.pytorch.org/t/fsdp-cudacachingallocator-an-outsider-newb-perspective/1486 cover this in details.

**Comment by awgu** [comment]
One stream per FSDP param group is a memory regression because those streams cannot share memory pools and FSDP does allocate memory in the reduce-scatter stream.

**Comment by kwen2501** [comment]
Outstanding FSDP instances cannot/should not share memory pool anyway, right? (because they will need to happen in parallel). 
Let's say the max outstanding is 2. Then we should have two streams, where FSDP instances 1, 3, 5 are on stream 1, and FSDP instances 2, 4, 6 are on stream 2. Then 1, 3, 5 should able to achieve reuse, and same for 2, 4, 6.

**Comment by awgu** [comment]
@kwen2501 there is only one outstanding reduce-scatter at once

**Comment by kwen2501** [comment]
That's true. However, when one instance's gradients go through reduce-scatter, doesn't the next instance's backward occur at the same time? If that's the case, wouldn't we 2x gradient buffer?

**Comment by awgu** [comment]
@kwen2501 I am not sure I understand your point. I agree that we will use a factor times the module's gradients worth of memory, but how are memory for reduce-scatter and memory for the next backward interchangeable? I was thinking you are proposing to reserve 2x instead of 1x memory for reduce-scatter.
